pkg/didcomm/messenger/messenger.go

```diff
@@ -0,0 +1,52 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package messenger
+
+import (
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/dispatcher"
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+// Provider contains dependencies for the Messenger
+type Provider interface {
+	OutboundDispatcher() dispatcher.Outbound
+	StorageProvider() storage.Provider
+}
+
+// Messenger describes the messenger structure
+type Messenger struct {
+	dispatcher dispatcher.Outbound
+}
+
+// NewMessenger returns a new instance of the Messenger
+func NewMessenger(ctx Provider) *Messenger {
+	return &Messenger{
+		dispatcher: ctx.OutboundDispatcher(),
+	}
+}
+
+// HandleInbound handles all inbound messages
+// function behavior is flexible and can be modified by providing options
+func (*Messenger) HandleInbound(msg service.DIDCommMsgMap, myDID, theirDID string, opts ...interface{}) error {
+	return nil
+}
+
+// Send sends the message by starting a new thread.
+func (m *Messenger) Send(msg service.DIDCommMsgMap, myDID, theirDID string) error {
```

Comment:
I'm confused on why the msg argument in these methods is not of the interface type `DIDCommMsg`.

I also see that calls deeper within SendToDID still use json.Marsal instead of DIDCommMsg.Encode.

Reply:
We do not need this interface here. The interface should be provided when you are using only read operation. Messenger aims to modify the message. If we will use interface we need to do type assertion under the hood. Which is bad in case if someone decided to provide a different type of message (custom implementation of did comm message interface).

We will use `unmarshal` for inbound messages and `marshal` for outbound messages (e.g see `ParseDIDCommMsgMap` function). I can create Encode function which will execute  `marshal` under the hood to convert message payload to bytes. But it will break current logic, it requires to change OutboundDispatcher send logic which is not simple.

Reply:
@troyronda @soluchok is this resolved?